src/pytorch_lightning/loops/optimization/optimizer_loop.py

```diff
@@ -142,91 +142,26 @@ def __call__(self, *args: Any, **kwargs: Any) -> Optional[Tensor]:
-    and the optimizer index if multiple optimizers are requested.
-
-    It is the leaf node in the tree of loops and performs automatic optimization
-    (forward, zero grad, backward, optimizer step).
-    """
+    """Performs automatic optimization (forward, zero grad, backward, optimizer step)"""
 
     output_result_cls = ClosureResult
 
     def __init__(self) -> None:
         super().__init__()
         self.optim_progress: OptimizationProgress = OptimizationProgress()
-
-        self._outputs: _OUTPUTS_TYPE = {}
         self._skip_backward: bool = False
-        self._optimizers: Tuple[Optimizer, ...] = tuple()
-        self._indices: Tuple[int, ...] = tuple()
-
-    @property
-    def optimizer_idx(self) -> int:
-        return self._indices[self.optim_progress.optimizer_position]
-
-    @property
-    def done(self) -> bool:
-        """Returns ``True`` when the last optimizer in the sequence has run."""
-        return self.optim_progress.optimizer_position >= len(self._indices)
-
-    def run(self, optimizers: List[Tuple[int, Optimizer]], kwargs: OrderedDict) -> _OUTPUTS_TYPE:
-        self.reset()
-        self.on_run_start(optimizers)
-        while not self.done:
-            try:
-                self.advance(kwargs)
-                self._restarting = False
-            except StopIteration:
-                break
-        self._restarting = False
-        return self.on_run_end()
-
-    def reset(self) -> None:
-        if not self.restarting:
-            # when reset() is called from outside (manually), we reset the loop progress
-            self.optim_progress.optimizer_position = 0
-        else:
-            self.optim_progress.reset_on_restart()
-        self._outputs = {}
-
-    def on_run_start(self, optimizers: List[Tuple[int, Optimizer]]) -> None:
-        self._indices, self._optimizers = zip(*optimizers)
-        if self.done:
-            self.optim_progress.optimizer_position = 0
-
-    def advance(self, kwargs: OrderedDict) -> None:
-        kwargs = self._build_kwargs(kwargs, self.optimizer_idx)
-
-        result = self._run_optimization(kwargs, self._optimizers[self.optim_progress.optimizer_position])
-        if result.loss is not None:
-            # automatic optimization assumes a loss needs to be returned for extras to be considered as the batch
-            # would be skipped otherwise
-            self._outputs[self.optimizer_idx] = result.asdict()
-        self.optim_progress.optimizer_position += 1
-
-    def on_run_end(self) -> _OUTPUTS_TYPE:
-        outputs, self._outputs = self._outputs, {}  # free memory
-        self._indices = tuple()
-        self._optimizers = tuple()
-        return outputs
-
-    def _run_optimization(self, kwargs: OrderedDict, optimizer: torch.optim.Optimizer) -> ClosureResult:
+
+    def run(self, optimizer: Optimizer, kwargs: OrderedDict) -> _OUTPUTS_TYPE:
```

Comment:
why do we need an entire loop for this if it isn't even looping? Can we either remove the `_Loop` baseclass as it is misleading or move this up in the loop hierarchy and have it as a simple function?

Reply:
Yes, this would be the direct follow up. I thought about doing it in this PR but it would only make it much bigger. There would be three follow up PRs:

- Remove optimizer loop as a loop class
- All documentation updates
- Clean up opt_idx in lightning_optimizer utility functions